src/transport/common/listener.rs

```diff
@@ -311,10 +367,27 @@ fn multiaddr_to_socket_address(
                 return Err(Error::AddressError(AddressError::InvalidProtocol));
             }
         },
-        Some(Protocol::Dns(address))
-        | Some(Protocol::Dns4(address))
-        | Some(Protocol::Dns6(address)) => match iter.next() {
-            Some(Protocol::Tcp(port)) => AddressType::Dns(address.to_string(), port),
+        Some(Protocol::Dns(address)) | Some(Protocol::Dns4(address)) => match iter.next() {
+            Some(Protocol::Tcp(port)) => AddressType::Dns {
+                address: address.into(),
+                port,
+                is_dns6: false,
```

Comment:
As [multiaddr spec](`https://github.com/multiformats/multiaddr/blob/master/protocols.csv`) says, `Protocol::Dns` can resolve to both IPv4 & IPv6. I would use this PR as an opportunity to fix the existing behavior and allow IPv6 addresses with `Protocol::Dns`.

Reply:
Nice catch! For sure, will fix it right away